File: fpv-receiver/src/stun.c
/*
 * STUN Client - Implementation
 */

#include "stun.h"
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <netdb.h>

#define STUN_BINDING_REQUEST 0x0001
#define STUN_BINDING_RESPONSE 0x0101
#define STUN_MAGIC_COOKIE 0x2112A442
#define ATTR_MAPPED_ADDRESS 0x0001
#define ATTR_XOR_MAPPED_ADDRESS 0x0020

static const char *stun_servers[] = {
    "stun.cloudflare.com",
    "stun.l.google.com",
    "stun1.l.google.com",
    NULL};

static int parse_stun_response(const uint8_t *buf, size_t len,
                               const uint8_t *txn_id,
                               struct sockaddr_in *public_addr)
{
    if (len < 20)
        return -1;

    uint16_t msg_type = (buf[0] << 8) | buf[1];
    if (msg_type != STUN_BINDING_RESPONSE)
        return -1;

    uint16_t msg_len = (buf[2] << 8) | buf[3];
    uint32_t magic = (buf[4] << 24) | (buf[5] << 16) | (buf[6] << 8) | buf[7];
    if (magic != STUN_MAGIC_COOKIE)
        return -1;

    /* Check transaction ID */
    if (memcmp(buf + 8, txn_id, 12) != 0)
        return -1;

    /* Parse attributes */
    size_t offset = 20;
    while (offset + 4 <= 20 + msg_len && offset + 4 <= len)
    {
        uint16_t attr_type = (buf[offset] << 8) | buf[offset + 1];
        uint16_t attr_len = (buf[offset + 2] << 8) | buf[offset + 3];

        if (offset + 4 + attr_len > len)
            break;

        const uint8_t *attr_data = buf + offset + 4;

        if (attr_type == ATTR_XOR_MAPPED_ADDRESS && attr_len >= 8)
        {
            uint8_t family = attr_data[1];
            if (family == 0x01)
            { /* IPv4 */
                uint16_t xport = (attr_data[2] << 8) | attr_data[3];
                uint16_t port = xport ^ (STUN_MAGIC_COOKIE >> 16);

                uint32_t xaddr = (attr_data[4] << 24) | (attr_data[5] << 16) |
                                 (attr_data[6] << 8) | attr_data[7];
                uint32_t addr = xaddr ^ STUN_MAGIC_COOKIE;

                public_addr->sin_family = AF_INET;
                public_addr->sin_port = htons(port);
                public_addr->sin_addr.s_addr = htonl(addr);
                return 0;
            }
        }
        else if (attr_type == ATTR_MAPPED_ADDRESS && attr_len >= 8)
        {
            uint8_t family = attr_data[1];
            if (family == 0x01)
            { /* IPv4 */
                uint16_t port = (attr_data[2] << 8) | attr_data[3];

                public_addr->sin_family = AF_INET;
                public_addr->sin_port = htons(port);
                memcpy(&public_addr->sin_addr.s_addr, attr_data + 4, 4);
                return 0;
            }
        }

        /* Next attribute (padded to 4 bytes) */
        offset += 4 + ((attr_len + 3) & ~3);
    }

    return -1;
}

int fpv_stun_discover(int sockfd, fpv_stun_result_t *result)
{
    /* Get local address */
    socklen_t addr_len = sizeof(result->local_addr);
    if (getsockname(sockfd, (struct sockaddr *)&result->local_addr, &addr_len) < 0)
    {
        return -1;
    }

    /* Build STUN binding request */
    uint8_t request[20];
    request[0] = (STUN_BINDING_REQUEST >> 8) & 0xFF;
    request[1] = STUN_BINDING_REQUEST & 0xFF;
    request[2] = 0; /* length = 0 */
    request[3] = 0;
    request[4] = (STUN_MAGIC_COOKIE >> 24) & 0xFF;
    request[5] = (STUN_MAGIC_COOKIE >> 16) & 0xFF;
    request[6] = (STUN_MAGIC_COOKIE >> 8) & 0xFF;
    request[7] = STUN_MAGIC_COOKIE & 0xFF;

    /* Transaction ID (12 random bytes) */
    uint8_t txn_id[12];
    for (int i = 0; i < 12; i++)
    {
        txn_id[i] = rand() & 0xFF;
    }
    memcpy(request + 8, txn_id, 12);

    /* Try each STUN server */
    for (int s = 0; stun_servers[s]; s++)
    {
        /* Resolve server */
        struct addrinfo hints = {0};
        hints.ai_family = AF_INET;
        hints.ai_socktype = SOCK_DGRAM;

        struct addrinfo *res;
        if (getaddrinfo(stun_servers[s], "3478", &hints, &res) != 0)
        {
            continue;
        }

        struct sockaddr_in server_addr;
        memcpy(&server_addr, res->ai_addr, sizeof(server_addr));
        freeaddrinfo(res);

        /* Try a few times */
        for (int attempt = 0; attempt < 3; attempt++)
        {
            /* Send request */
            if (sendto(sockfd, request, sizeof(request), 0,
                       (struct sockaddr *)&server_addr, sizeof(server_addr)) < 0)
            {
                continue;
            }

            /* Wait for response (with timeout) */
            fd_set fds;
            FD_ZERO(&fds);
            FD_SET(sockfd, &fds);

            struct timeval tv = {.tv_sec = 1, .tv_usec = 0};
            int ready = select(sockfd + 1, &fds, NULL, NULL, &tv);

            if (ready <= 0)
                continue;

            /* Receive response */
            uint8_t response[1024];
            ssize_t n = recv(sockfd, response, sizeof(response), 0);
            if (n < 20)
                continue;

            /* Parse response */
            if (parse_stun_response(response, n, txn_id, &result->public_addr) == 0)
            {
                strncpy(result->server, stun_servers[s], sizeof(result->server) - 1);
                return 0;
            }
        }
    }

    return -1;
}

File: fpv-sender/stun/stun.go
// Package stun implements a minimal STUN client for NAT discovery (RFC 5389).
package stun

import (
	"context"
	"crypto/rand"
	"encoding/binary"
	"errors"
	"fmt"
	"net"
	"time"
)

// STUN message types
const (
	bindingRequest  = 0x0001
	bindingResponse = 0x0101
)

// STUN attributes
const (
	attrMappedAddress    = 0x0001
	attrXorMappedAddress = 0x0020
)

// STUN magic cookie (RFC 5389)
const magicCookie = 0x2112A442

// Default STUN servers
var DefaultServers = []string{
	"stun.cloudflare.com:3478",
	"stun.l.google.com:19302",
	"stun1.l.google.com:19302",
}

// Errors
var (
	ErrNoResponse   = errors.New("no STUN response received")
	ErrInvalidReply = errors.New("invalid STUN reply")
)

// Result contains the discovered endpoints.
type Result struct {
	LocalAddr  *net.UDPAddr // Local bound address
	PublicAddr *net.UDPAddr // Server-reflexive (public) address
	Server     string       // STUN server that responded
}

// Discover performs STUN binding to discover the public endpoint.
// It uses the provided UDP connection (which should already be bound).
func Discover(ctx context.Context, conn *net.UDPConn, servers []string) (*Result, error) {
	if servers == nil {
		servers = DefaultServers
	}

	localAddr := conn.LocalAddr().(*net.UDPAddr)

	for _, server := range servers {
		serverAddr, err := net.ResolveUDPAddr("udp4", server)
		if err != nil {
			continue
		}

		publicAddr, err := doBinding(ctx, conn, serverAddr)
		if err != nil {
			continue
		}

		return &Result{
			LocalAddr:  localAddr,
			PublicAddr: publicAddr,
			Server:     server,
		}, nil
	}

	return nil, ErrNoResponse
}

// doBinding sends a STUN binding request and waits for response.
func doBinding(ctx context.Context, conn *net.UDPConn, server *net.UDPAddr) (*net.UDPAddr, error) {
	// Generate transaction ID (12 bytes)
	txnID := make([]byte, 12)
	if _, err := rand.Read(txnID); err != nil {
		return nil, err
	}

	// Build request: type (2) + length (2) + magic cookie (4) + txn_id (12) = 20 bytes
	req := make([]byte, 20)
	binary.BigEndian.PutUint16(req[0:2], bindingRequest)
	binary.BigEndian.PutUint16(req[2:4], 0) // length = 0 (no attributes)
	binary.BigEndian.PutUint32(req[4:8], magicCookie)
	copy(req[8:20], txnID)

	// Set read deadline
	deadline, ok := ctx.Deadline()
	if !ok {
		deadline = time.Now().Add(2 * time.Second)
	}
	conn.SetReadDeadline(deadline)

	// Send request (with retries)
	for attempt := 0; attempt < 3; attempt++ {
		if _, err := conn.WriteToUDP(req, server); err != nil {
			return nil, err
		}

		// Wait for response
		buf := make([]byte, 1024)
		n, _, err := conn.ReadFromUDP(buf)
		if err != nil {
			if nerr, ok := err.(net.Error); ok && nerr.Timeout() {
				continue // Retry on timeout
			}
			return nil, err
		}

		// Parse response
		addr, err := parseResponse(buf[:n], txnID)
		if err != nil {
			continue
		}
		return addr, nil
	}

	return nil, ErrNoResponse
}

// parseResponse parses a STUN binding response.
func parseResponse(buf []byte, expectedTxnID []byte) (*net.UDPAddr, error) {
	if len(buf) < 20 {
		return nil, ErrInvalidReply
	}

	msgType := binary.BigEndian.Uint16(buf[0:2])
	if msgType != bindingResponse {
		return nil, ErrInvalidReply
	}

	msgLen := binary.BigEndian.Uint16(buf[2:4])
	magic := binary.BigEndian.Uint32(buf[4:8])
	if magic != magicCookie {
		return nil, ErrInvalidReply
	}

	// Check transaction ID
	if len(buf) < 20 {
		return nil, ErrInvalidReply
	}
	for i := 0; i < 12; i++ {
		if buf[8+i] != expectedTxnID[i] {
			return nil, ErrInvalidReply
		}
	}

	// Parse attributes
	offset := 20
	end := 20 + int(msgLen)
	if end > len(buf) {
		return nil, ErrInvalidReply
	}

	for offset+4 <= end {
		attrType := binary.BigEndian.Uint16(buf[offset : offset+2])
		attrLen := binary.BigEndian.Uint16(buf[offset+2 : offset+4])
		attrData := buf[offset+4 : offset+4+int(attrLen)]

		if attrType == attrXorMappedAddress && attrLen >= 8 {
			// XOR-MAPPED-ADDRESS (preferred)
			family := attrData[1]
			if family == 0x01 { // IPv4
				xport := binary.BigEndian.Uint16(attrData[2:4])
				port := xport ^ uint16(magicCookie>>16)

				xaddr := binary.BigEndian.Uint32(attrData[4:8])
				addr := xaddr ^ magicCookie

				ip := net.IPv4(
					byte(addr>>24),
					byte(addr>>16),
					byte(addr>>8),
					byte(addr),
				)
				return &net.UDPAddr{IP: ip, Port: int(port)}, nil
			}
		} else if attrType == attrMappedAddress && attrLen >= 8 {
			// MAPPED-ADDRESS (fallback)
			family := attrData[1]
			if family == 0x01 { // IPv4
				port := binary.BigEndian.Uint16(attrData[2:4])
				ip := net.IPv4(attrData[4], attrData[5], attrData[6], attrData[7])
				return &net.UDPAddr{IP: ip, Port: int(port)}, nil
			}
		}

		// Move to next attribute (padded to 4-byte boundary)
		offset += 4 + int((attrLen+3) & ^uint16(3))
	}

	return nil, ErrInvalidReply
}

// DiscoverWithNewSocket creates a new UDP socket, performs STUN, and returns both.
func DiscoverWithNewSocket(ctx context.Context, localPort int) (*net.UDPConn, *Result, error) {
	addr := &net.UDPAddr{IP: net.IPv4zero, Port: localPort}
	conn, err := net.ListenUDP("udp4", addr)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to create UDP socket: %w", err)
	}

	result, err := Discover(ctx, conn, nil)
	if err != nil {
		conn.Close()
		return nil, nil, err
	}

	return conn, result, nil
}
